services/galley/src/Galley/API/Action.hs

```diff
@@ -401,38 +401,50 @@ instance IsConversationAction ConversationAccessData where
       traverse_ (E.deleteBot (tUnqualified lcnv) . botMemId) (bmBots toRemove)
 
       -- Update current bots and members
-      let current' = current {bmBots = bmBots desired}
+      -- current bots and members but only desired bots
+      let bmToNotify = current {bmBots = bmBots desired}
 
       -- Remove users and notify everyone
       void . for_ (nonEmpty (bmQualifiedMembers lcnv toRemove)) $ \usersToRemove -> do
         let rAction = ConversationLeave usersToRemove
         void . runError @NoChanges $ performAction qusr lcnv conv rAction
-        notifyConversationAction qusr Nothing lcnv current' (conversationAction rAction)
+        notifyConversationAction qusr Nothing lcnv bmToNotify (conversationAction rAction)
     pure (mempty, action)
     where
-      filterActivated :: Member BrigAccess r => BotsAndMembers -> Sem r BotsAndMembers
-      filterActivated bm
-        | convAccessRole conv > ActivatedAccessRole
-            && cupAccessRole action <= ActivatedAccessRole = do
-          activated <- map User.userId <$> E.lookupActivatedUsers (toList (bmLocals bm))
-          -- FUTUREWORK: should we also remove non-activated remote users?
-          pure $ bm {bmLocals = Set.fromList activated}
-        | otherwise = pure bm
-
-      filterTeammates :: Member TeamStore r => BotsAndMembers -> Sem r BotsAndMembers
-      filterTeammates bm = do
-        -- In a team-only conversation we also want to remove bots and guests
-        case (cupAccessRole action, convTeam conv) of
-          (TeamAccessRole, Just tid) -> do
-            onlyTeamUsers <- flip filterM (toList (bmLocals bm)) $ \user ->
-              isJust <$> E.getTeamMember tid user
-            pure $
-              BotsAndMembers
-                { bmLocals = Set.fromList onlyTeamUsers,
-                  bmBots = mempty,
-                  bmRemotes = mempty
-                }
-          _ -> pure bm
+      maybeRemoveBots :: Member BrigAccess r => BotsAndMembers -> Sem r BotsAndMembers
+      maybeRemoveBots bm =
+        if Set.member ServiceAccessRole (cupAccessRoles action)
+          then pure bm
+          else pure $ bm {bmBots = mempty}
+
+      maybeRemoveGuests :: Member BrigAccess r => BotsAndMembers -> Sem r BotsAndMembers
+      maybeRemoveGuests bm =
+        if Set.member GuestAccessRole (cupAccessRoles action)
+          then pure bm
+          else do
+            activated <- map User.userId <$> E.lookupActivatedUsers (toList (bmLocals bm))
+            -- FUTUREWORK: should we also remove non-activated remote users?
```

Comment:
Federation people: can users joining via a guest link without account be non-local?